src/Microsoft.Unity.Analyzers/ImproperSerializeField.cs

```diff
@@ -0,0 +1,183 @@
+			if (!containingType.Extends(typeof(UnityEngine.Object)))
+				return false;
+
+			bool hasSerializeFieldAttribute = symbol
+				.GetAttributes()
+				.Any(a => a.AttributeClass.Matches(typeof(UnityEngine.SerializeField)));
+
+			if (!hasSerializeFieldAttribute)
+				return false;
+
+			propertyName = symbol.Name;
+			return true;
+		}
+
+		private static bool FieldHasInvalidSerializeField(List<ISymbol> symbols, out string fields)
+		{
+			fields = null;
+			List<string> fieldNames = new List<string>();
+
+			bool foundMatchingSymbol = false;
+			foreach (var symbol in symbols)
+			{
+				fieldNames.Add(symbol.Name);
+
+				if (!(symbol is IFieldSymbol && symbol.DeclaredAccessibility == Accessibility.Public))
+					continue;
+
+				var containingType = symbol.ContainingType;
+				if (!containingType.Extends(typeof(UnityEngine.Object)))
+					continue;
+
+				if (!symbol.GetAttributes().Any(a => a.AttributeClass.Matches(typeof(UnityEngine.SerializeField))))
+					continue;
+
+				foundMatchingSymbol = true;
+			}
+
+			fields = string.Join(", ", fieldNames.ToArray());
+			return foundMatchingSymbol;
+		}
+	}
+
+	[ExportCodeFixProvider(LanguageNames.CSharp)]
+	public class ImproperSerializeFieldCodeFix : CodeFixProvider
+	{
+		public sealed override ImmutableArray<string> FixableDiagnosticIds => ImmutableArray.Create(ImproperSerializeFieldAnalyzer.Id);
+
+		public sealed override FixAllProvider GetFixAllProvider() => WellKnownFixAllProviders.BatchFixer;
+
+		public sealed override async Task RegisterCodeFixesAsync(CodeFixContext context)
+		{
+			var root = await context.Document.GetSyntaxRootAsync(context.CancellationToken).ConfigureAwait(false);
+
+			var declaration = root.FindNode(context.Span) as MemberDeclarationSyntax;
+			if (declaration == null)
+				return;
+
+			if (!(declaration is PropertyDeclarationSyntax || declaration is FieldDeclarationSyntax))
+				return;
+
+			context.RegisterCodeFix(
+				CodeAction.Create(
+					Strings.ImproperSerializeFieldCodeFixTitle,
+					ct => RemoveSerializeFieldAttribute(context.Document, context, declaration, ct),
+					declaration.ToFullString()),
+				context.Diagnostics);
+		}
+
+		private static async Task<Document> RemoveSerializeFieldAttribute(Document document, CodeFixContext context, MemberDeclarationSyntax declaration, CancellationToken cancellationToken)
+		{
+			var root = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+			var model = await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+
+			var attributes = new SyntaxList<AttributeListSyntax>();
+
+			foreach (var attributeList in declaration.AttributeLists)
+			{
+				List<AttributeSyntax> nodes = new List<AttributeSyntax>();
+				foreach (var node in attributeList.Attributes)
+				{
+					var attributeType = model.GetTypeInfo(node);
+					if (attributeType.Type.Matches(typeof(UnityEngine.SerializeField)))
+						nodes.Add(node);
+				}
+
+				if (nodes.Count() > 0)
```

Comment:
Perhaps we can remove the block and use to reduce nesting:
```
				if (!nodes.Any()) 
					continue;
```

Reply:
There's now an Else here, so not exactly, but I've incorporated other parts of this feedback into an upcoming commit.